### PR TITLE
man: explicitly note that ExecSt*Post does count for After/Before ordering

### PR DESCRIPTION
**Pull request by bluca**
It was not obvious to me and a few others, so spell it out explicitly.

### CODE REVIEW

**Comment by bluca** [comment]
semaphoreci is complaining about an udev test, so I'll assume it's unrelated to man pages changes:

```
-- stderr ---
umount: test/tmpfs: no mount point specified.
/run/udev/rules.d/udev-test.rules:1 Unknown user 'bad', ignoring
/run/udev/rules.d/udev-test.rules:1 Unknown group 'name', ignoring
/run/udev/rules.d/udev-test.rules:1 Unknown user 'bad', ignoring
/run/udev/rules.d/udev-test.rules:1 Unknown group 'name', ignoring
ttyACM0: Conflicting device node '/dev/ttyACM0' found, link to '/dev/ttyACM0' will not be created.
ttyACM0: Conflicting device node '/dev/ttyACM0' found, link to '/dev/ttyACM0' will not be created.
ttyACM0: Conflicting device node '/dev/ttyACM0' found, link to '/dev/ttyACM0' will not be created.
ttyACM0: Conflicting device node '/dev/ttyACM0' found, link to '/dev/ttyACM0' will not be created.
ttyACM0: Conflicting device node '/dev/ttyACM0' found, link to '/dev/ttyACM0' will not be created.
ttyACM0: Conflicting device node '/dev/ttyACM0' found, link to '/dev/ttyACM0' will not be created.
ttyACM0: Conflicting device node '/dev/ttyACM0' found, link to '/dev/ttyACM0' will not be created.
sda: Failed to rename '/dev/..tmp-b8:0' to '/dev/.': Device or resource busy
tty0: Conflicting device node '/dev/tty0' found, link to '/dev/tty0' will not be created.
ttyACM0: Conflicting device node '/dev/ttyACM0' found, link to '/dev/ttyACM0' will not be created.
ttyACM0: Conflicting device node '/dev/ttyACM0' found, link to '/dev/ttyACM0' will not be created.
ttyACM0: Conflicting device node '/dev/ttyACM0' found, link to '/dev/ttyACM0' will not be created.
Can't exec "tree": No such file or directory at /tmp/autopkgtest-lxc.y9nqvscx/downtmp/build.doH/src/test/udev-test.pl line 1741.
/run/udev/rules.d/udev-test.rules:1 Invalid value for OPTIONS key, ignoring: 'last_rule'
/run/udev/rules.d/udev-test.rules:1 Invalid value for OPTIONS key, ignoring: 'last_rule'
/run/udev/rules.d/udev-test.rules:2 Unknown group 'not-ok', ignoring
/run/udev/rules.d/udev-test.rules:2 Unknown group 'not-ok', ignoring
/run/udev/rules.d/udev-test.rules:3 Unknown group 'not-ok', ignoring
/run/udev/rules.d/udev-test.rules:3 Unknown group 'not-ok', ignoring
/run/udev/rules.d/udev-test.rules:1 Invalid value "" for TAG (char 0: empty value), ignoring, but please fix it.
/run/udev/rules.d/udev-test.rules:1 Invalid value "" for TAG (char 0: empty value), ignoring, but please fix it.
/run/udev/rules.d/udev-test.rules:1 Invalid value "" for TAG (char 0: empty value), ignoring, but please fix it.
/run/udev/rules.d/udev-test.rules:1 Invalid value "" for TAG (char 0: empty value), ignoring, but please fix it.
/run/udev/rules.d/udev-test.rules:1 Invalid value "" for TAG (char 0: empty value), ignoring, but please fix it.
/run/udev/rules.d/udev-test.rules:1 Invalid value "" for TAG (char 0: empty value), ignoring, but please fix it.
/run/udev/rules.d/udev-test.rules:1: GOTO="does-not-exist" has no matching label, ignoring
/run/udev/rules.d/udev-test.rules:1: The line takes no effect any more, dropping
/run/udev/rules.d/udev-test.rules:1: GOTO="does-not-exist" has no matching label, ignoring
/run/udev/rules.d/udev-test.rules:1: The line takes no effect any more, dropping
/run/udev/rules.d/udev-test.rules:1 Invalid operator for KERNEL.
/run/udev/rules.d/udev-test.rules:1 Invalid operator for KERNEL.
/run/udev/rules.d/udev-test.rules:3 Invalid key/value pair, ignoring.
/run/udev/rules.d/udev-test.rules:3 Invalid key/value pair, ignoring.
sda: /run/udev/rules.d/udev-test.rules:2 Failed to parse mode '0400letsdoabuffferoverflow0123456789012345789012345678901234567890', ignoring: Invalid argument
sda: /run/udev/rules.d/udev-test.rules:2 Failed to parse mode '0400letsdoabuffferoverflow0123456789012345789012345678901234567890', ignoring: Invalid argument
/run/udev/rules.d/udev-test.rules:10001: Line is too long, ignored
/run/udev/rules.d/udev-test.rules:10001: Line is too long, ignored
```

**Comment by boucman** [comment]
you sure about that ? i'm pretty sure it did, that's more or less the whole point of ExecStartPost... doing something after the ExecStart is ready but before anybody else has started using it...

**Comment by bluca** [comment]
> you sure about that ? i'm pretty sure it did, that's more or less the whole point of ExecStartPost... doing something after the ExecStart is ready but before anybody else has started using it...

From \#systemd on IRC:

> (15:06:46) bluca: in the documentation for After/Before:
> (15:07:05) bluca: "Most importantly, for service units start-up is considered completed for the purpose of Before=/After= when all its configured start-up commands have been invoked and they either failed or reported start-up success."
> (15:07:19) bluca: 1) does that include ExecStartPOST ?
> (15:07:32) bluca: 2) what about ExecStopPost for the shutdown case?
> (15:10:24) bluca: I had assumed yes to both, but seeing something strange..
> (15:20:45) dreisner: bluca: no, ExecStartPost is run after the service is healthy. "ExecStartPost= commands are only run after the commands specified in ExecStart= have been invoked successfully, as determined by Type=" from the manpage
> (15:21:36) dreisner: or "ready", because systemd really has no concept of "healthy" aside from it's weird watchdog stuff

**Comment by boucman** [comment]
ExecStartPost is run after ExecStart has reported that it is healthy, yes... but the unit as a whole is still in the activating state until all ExecStartPost have finished running and that's what matter for Before/After....

I'm pretty sure of what I'm saying, it's even one of the test we do in my systemd course to show the purpose of ExecStartPost... introducing a delay and seeing that it does indeed delay the startup of dependencies...

Did you check that yourself ?

**Comment by keszybz** [comment]
Hmm,
```console
$ systemctl --user cat slowstart slowstart2 
# /home/fedora/.config/systemd/user/slowstart.service
[Service]
Type=oneshot
ExecStart=sleep 3
ExecStartPost=sleep 3

# /home/fedora/.config/systemd/user/slowstart2.service
[Unit]
Wants=slowstart.service
After=slowstart.service

[Service]
Type=oneshot
ExecStart=sleep 3
$ time systemctl --user start slowstart2 

real	0m9.491s
user	0m0.010s
sys	0m0.017s
```

It seems to be "taken into account", at least in the Type=oneshot case.

**Comment by bluca** [comment]
> Did you check that yourself ?

I did - but, as it turns out, it was something else going wrong, unrelated (a transient unit not being removed after it's stopped), so it was a complete red herring. I trusted the comment on IRC as it was a very simple and convenient explanation for what I was seeing.

Thank you both for chiming in.

**Comment by bluca** [comment]
While I have the PR here, maybe it's worth to explicitly note that they count? So that there's no room for misinterpretations.

**Comment by keszybz** [comment]
LGTM.

File: man/systemd.unit.xml [hunk 1]
```diff
@@ -750,7 +750,7 @@
         type when precisely a unit has finished starting up. Most importantly, for service units start-up is
         considered completed for the purpose of <varname>Before=</varname>/<varname>After=</varname> when all
         its configured start-up commands have been invoked and they either failed or reported start-up
-        success.</para>
+        success. Note that this does includes ExecStartPost (or ExecStopPost for the shutdown case).</para>
```
Comment (poettering) [inline man/systemd.unit.xml:753]: enclose the settings in `<varname>` and suffix with `=` please

Comment (bluca) [reply man/systemd.unit.xml:753]: Whops, my bad - follow-up: https://github.com/systemd/systemd/pull/15374

**Pull request merged**